### PR TITLE
[CoreFoundation, ObjCRuntime] Add DispatchBlock APIs, in particular those that surface QOS

### PR DESCRIPTION
**Pull request by migueldeicaza**

### CODE REVIEW

File: src/CoreFoundation/DispatchBlock.cs [hunk 1]
```diff
@@ -15,8 +15,110 @@
...
+		UserInteractive = 0x21,
+		UserInitiated = 0x19,
+		Default = 0x15,
+		Utility = 0x11,
+		Background = 0x09,
+		Unspecified = 0
+	}
+
+	[StructLayout (LayoutKind.Sequential)]
+	internal struct DispatchBlock {
```
Comment (Therzok) [inline src/CoreFoundation/DispatchBlock.cs]:
```suggestion
	internal readonly struct DispatchBlock {
```

File: src/CoreFoundation/DispatchBlock.cs [hunk 2]
```diff
@@ -15,8 +15,110 @@
...
+		IntPtr block;
```
Comment (Therzok) [inline src/CoreFoundation/DispatchBlock.cs]:
```suggestion
		readonly IntPtr block;
```

File: src/CoreFoundation/DispatchBlock.cs [hunk 3]
```diff
@@ -15,8 +15,110 @@
...
+		UserInteractive = 0x21,
+		UserInitiated = 0x19,
+		Default = 0x15,
+		Utility = 0x11,
+		Background = 0x09,
+		Unspecified = 0
+	}
+
+	[StructLayout (LayoutKind.Sequential)]
+	internal readonly struct DispatchBlock {
```
[the remark this answers is not in the document]
Comment (Therzok) [reply src/CoreFoundation/DispatchBlock.cs]: That was my suggestion, sorry.

Can we make it so we pass LangVersion 7.2 to csc/msbuild?

[the remark this answers is not in the document]
Comment (rolfbjarne) [reply src/CoreFoundation/DispatchBlock.cs]: > Can we make it so we pass LangVersion 7.2 to csc/msbuild?

We could, but it's really unrelated to this PR, and I'd prefer it if we didn't mix things in PRs.

File: src/CoreFoundation/DispatchBlock.cs [hunk 4]
```diff
@@ -15,8 +15,107 @@
...
+		[DllImport (Constants.libcLibrary)]
+		extern static long dispatch_block_testcancel(IntPtr block);
+
+		public long TestCancel ()
+		{
+			return dispatch_block_testcancel (blockHandle);
+		}
+
+		[DllImport (Constants.libcLibrary)]
+		extern static long dispatch_block_wait (IntPtr block, DispatchTime time);
```
Comment (spouliot) [inline src/CoreFoundation/DispatchBlock.cs]: it should return `nint`

File: src/CoreFoundation/DispatchBlock.cs [hunk 5]
```diff
@@ -15,8 +15,107 @@
...
+
+		public void Wait (DispatchTime time)
+		{
+			dispatch_block_wait (blockHandle, time);
```
Comment (spouliot) [inline src/CoreFoundation/DispatchBlock.cs]: this hides the return value (i.e. it should also return `nint`)

File: src/CoreFoundation/DispatchBlock.cs [hunk 6]
```diff
@@ -15,8 +15,107 @@
...
+				throw new ArgumentNullException (nameof (queue));
+			if (notification == null)
+				throw new ArgumentNullException (nameof (notification));
+			dispatch_block_notify (blockHandle, queue.Handle, notification.blockHandle);
+		}
+		
+		[DllImport (Constants.libcLibrary)]
+		extern static long dispatch_block_testcancel(IntPtr block);
+
+		public long TestCancel ()
```
Comment (spouliot) [inline src/CoreFoundation/DispatchBlock.cs]: no reason for not returning `nint`

File: src/ObjCRuntime/Blocks.cs [hunk 7]
```diff
@@ -341,7 +341,7 @@ internal static IntPtr GetBlockForDelegate (MethodInfo minfo, object @delegate,
 		}
 
 		[DllImport (Messaging.LIBOBJC_DYLIB)]
-		static extern IntPtr _Block_copy (ref BlockLiteral block);
+		internal static extern IntPtr _Block_copy (ref BlockLiteral block);
```
Comment (spouliot) [inline src/ObjCRuntime/Blocks.cs]: needed ? it's already (re)declared in `DispatchBlock` ?

File: src/CoreFoundation/DispatchBlock.cs [hunk 8]
```diff
@@ -33,6 +132,35 @@ public static unsafe void Invoke (Action codeToRun, Action<IntPtr> invoker)
...
+		{
+			if (blockHandle != IntPtr.Zero){
+				_Block_release (blockHandle);
+				blockHandle = IntPtr.Zero;
+			}
+		}
+	}
+
+	[Flags]
+	public enum DispatchBlockFlags : ulong {
```
[the remark this answers is not in the document]
Comment (rolfbjarne) [reply src/CoreFoundation/DispatchBlock.cs]: Fixed.

File: src/ObjCRuntime/Blocks.cs (hunk 7 above)
Comment (rolfbjarne) [reply src/ObjCRuntime/Blocks.cs]: I've moved all the P/Invokes in DispatchBlock here.

File: src/CoreFoundation/DispatchBlock.cs (hunk 6 above)
Comment (rolfbjarne) [reply src/CoreFoundation/DispatchBlock.cs]: Fixed.

File: src/CoreFoundation/DispatchBlock.cs (hunk 5 above)
Comment (rolfbjarne) [reply src/CoreFoundation/DispatchBlock.cs]: Fixed.

File: src/CoreFoundation/DispatchBlock.cs (hunk 4 above)
Comment (rolfbjarne) [reply src/CoreFoundation/DispatchBlock.cs]: Fixed.

File: src/CoreFoundation/DispatchBlock.cs [hunk 9]
```diff
@@ -15,8 +15,107 @@
...
+		{
+			if (queue == null)
+				throw new ArgumentNullException (nameof (queue));
+			if (notification == null)
+				throw new ArgumentNullException (nameof (notification));
+			dispatch_block_notify (blockHandle, queue.Handle, notification.blockHandle);
+		}
+		
+		[DllImport (Constants.libcLibrary)]
+		extern static long dispatch_block_testcancel(IntPtr block);
```
[the remark this answers is not in the document]
Comment (rolfbjarne) [reply src/CoreFoundation/DispatchBlock.cs]: Fixed.

File: src/CoreFoundation/DispatchBlock.cs [hunk 10]
```diff
@@ -15,8 +15,107 @@
...
+			        BlockLiteral *block_ptr_handler = &block_handler;
+			        block_handler.SetupBlockUnsafe (BlockStaticDispatchClass.static_dispatch_block, action);
+				var ret =  _Block_copy (block_ptr_handler);
+				block_handler.CleanupBlock ();
+				return new DispatchBlock (dispatch_block_create (flags, ret));
+			}
+		}
+
+		[DllImport (Constants.libcLibrary)]
+		extern static IntPtr dispatch_block_create_with_qos_class (DispatchBlockFlags flags, DispatchQualityOfService qosClass, int relative_priority, IntPtr dispatchBlock);
```
[the remark this answers is not in the document]
Comment (rolfbjarne) [reply src/CoreFoundation/DispatchBlock.cs]: Fixed.

File: src/CoreFoundation/DispatchBlock.cs [hunk 11]
```diff
@@ -15,8 +15,107 @@
...
+			        BlockLiteral *block_ptr_handler = &block_handler;
+			        block_handler.SetupBlockUnsafe (BlockStaticDispatchClass.static_dispatch_block, action);
+				DispatchBlock ret = new DispatchBlock (_Block_copy (block_ptr_handler));
+				block_handler.CleanupBlock ();
+				return ret;
+			}
+		}
+
+		[DllImport (Constants.libcLibrary)]
+		extern static IntPtr dispatch_block_create (DispatchBlockFlags flags, IntPtr block);
```
[the remark this answers is not in the document]
Comment (rolfbjarne) [reply src/CoreFoundation/DispatchBlock.cs]: Fixed.

**Comment by rolfbjarne** [comment]
Test failures are both unrelated (https://github.com/xamarin/maccore/issues/581).

@migueldeicaza does the current PR look good to you?

**Comment by migueldeicaza** [comment]
This is a thing of beauty, and thanks for catching that long was not long, but nint.

Can we merge?

**Comment by lewurm** [comment]
I get this failure on 64bit devices:
```
[FAIL] DispatchBlockTest.ExplicitActionConversionInvoke : System.ExecutionEngineException : Attempting to JIT compile method '(wrapper managed-to-native) CoreFoundation.DispatchBlock/DispatchBlockCallback:wrapper_aot_native (object,intptr)' while running in aot-only mode. See https://docs.microsoft.com/xamarin/ios/internals/limitations for more information. 
[FAIL] DispatchBlockTest.Invoke : System.ExecutionEngineException : Attempting to JIT compile method '(wrapper managed-to-native) CoreFoundation.DispatchBlock/DispatchBlockCallback:wrapper_aot_native (object,intptr)' while running in aot-only mode. See https://docs.microsoft.com/xamarin/ios/internals/limitations for more information. 
```

**Comment by rolfbjarne** [comment]
> I get this failure on 64bit devices:

I'll this fix.

**Pull request merged**